Error in pip install duckDB

### What happens?

Afternoon guys, 

to start, massive fan of duckDB! This stuff is transformational. 

I'm getting an error when i tried to install duckDB (pip install duckdb==0.9.0). 

I'm installing on a venv, on python 3.12.0. This is a brand new server i'm having issue with, i have no issue with other server running older version (in fact, i could actually upgrade it). 

I can't debug the error message and i'm getting below (a snapshot of a long one). 

      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(267): error C2485: '__restrict': unrecognized extended attribute
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(269): error C2485: '__restrict': unrecognized extended attribute
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(273): error C2485: '__restrict': unrecognized extended attribute
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(275): error C2485: '__restrict': unrecognized extended attribute
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(279): error C2485: '__restrict': unrecognized extended attribute
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(281): error C2485: '__restrict': unrecognized extended attribute
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/internal/jemalloc_internal_macros.h(36): warning C4005: 'restrict': macro redefinition
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/internal/jemalloc_internal_decls.h(87): note: see previous definition of 'restrict'
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.37.32822\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for duckdb
Failed to build duckdb
ERROR: Could not build wheels for duckdb, which is required to install pyproject.toml-based projects

### To Reproduce

pip install duckdb==0.9.0

### OS:

x64

### DuckDB Version:

0.9.0

### DuckDB Client:

Python

### Full Name:

Maha Kepakisan

### Affiliation:

DDB

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

Hello guys. I am getting a similar error in my laptop. 

It's a samsung with Microsoft Windows 11 Home.
Processor: 12th Gen Intel(R) Core(TM) i7-1260P   2.10 GHz

Python version: 3.12.0

## To Reproduce
pip install duckdb (I tried with version 0.9.0 and 0.9.1)

If you need more information let me know
```
"C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.37.32822\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -Iduckdb_build/src/include -Iduckdb_build/third_party/fmt/include -Iduckdb_build/third_party/fsst -Iduckdb_build/third_party/re2 -Iduckdb_build/third_party/miniz -Iduckdb_build/third_party/utf8proc/include -Iduckdb_build/third_party/utf8proc -Iduckdb_build/third_party/hyperloglog -Iduckdb_build/third_party/fastpforlib -Iduckdb_build/third_party/tdigest -Iduckdb_build/third_party/libpg_query/include -Iduckdb_build/third_party/libpg_query -Iduckdb_build/third_party/concurrentqueue -Iduckdb_build/third_party/pcg -Iduckdb_build/third_party/httplib -Iduckdb_build/third_party/fast_float -Iduckdb_build/third_party/mbedtls -Iduckdb_build/third_party/mbedtls/include -Iduckdb_build/third_party/mbedtls/library -Iduckdb_build/third_party/jaro_winkler -Iduckdb_build/third_party/jaro_winkler/details -Iduckdb_build/extension/parquet/include -Iduckdb_build/third_party/parquet -Iduckdb_build/third_party/snappy -Iduckdb_build/third_party/thrift -Iduckdb_build/third_party/zstd/include -Iduckdb_build/extension/icu/include -Iduckdb_build/extension/icu/third_party/icu/common -Iduckdb_build/extension/icu/third_party/icu/i18n -Iduckdb_build/extension/fts/include -Iduckdb_build/third_party/snowball/libstemmer -Iduckdb_build/third_party/snowball/runtime -Iduckdb_build/third_party/snowball/src_c -Iduckdb_build/extension/tpch/include -Iduckdb_build/extension/tpch/dbgen/include -Iduckdb_build/extension/tpcds/include -Iduckdb_build/extension/tpcds/dsdgen/include -Iduckdb_build/extension/tpcds/dsdgen/include/dsdgen-c -Iduckdb_build/extension/json/include -Iduckdb_build/extension/json/yyjson/include -Iduckdb_build/extension/jemalloc/include -Iduckdb_build/extension/jemalloc/jemalloc/include -Iduckdb_build -IC:\Users\marti\AppData\Local\Temp\pip-install-84hl8hgp\duckdb_d22e3b7f4a624e358927192e2ecfe0a2\src\include -IC:\Users\marti\AppData\Local\Temp\pip-build-env-t7eu2gek\normal\Lib\site-packages\pybind11\include -IC:\Users\marti\AppData\Local\Temp\pip-build-env-t7eu2gek\normal\Lib\site-packages\pybind11\include -IC:\Users\marti\AppData\Local\Programs\Python\Python312\include -IC:\Users\marti\AppData\Local\Programs\Python\Python312\Include "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.37.32822\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\shared" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\winrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\cppwinrt" /EHsc /Tpduckdb_build/extension/fts/fts_extension.cpp /Fobuild\temp.win-amd64-cpython-312\Release\duckdb_build/extension/fts/fts_extension.obj /wd4244 /wd4267 /wd4200 /wd26451 /wd26495 /D_CRT_SECURE_NO_WARNINGS /utf-8 -DDUCKDB_PYTHON_LIB_NAME=duckdb -DDUCKDB_BUILD_LIBRARY -DWIN32 -DDUCKDB_EXTENSION_PARQUET_LINKED -DDUCKDB_EXTENSION_ICU_LINKED -DDUCKDB_EXTENSION_FTS_LINKED -DDUCKDB_EXTENSION_TPCH_LINKED -DDUCKDB_EXTENSION_JSON_LINKED -DDUCKDB_EXTENSION_AUTOLOAD_DEFAULT=1 -DDUCKDB_EXTENSION_AUTOINSTALL_DEFAULT=1
# the line above repeats in between these .cpp files
      fts_extension.cpp
      fts_indexing.cpp
      icu-dateadd.cpp
      icu-datefunc.cpp
      icu-datepart.cpp
      icu-datesub.cpp
      icu-datetrunc.cpp
      icu-list-range.cpp
      icu-makedate.cpp
      icu-strptime.cpp
      icu-table-range.cpp
      icu-timebucket.cpp
      icu-timezone.cpp
      icu_extension.cpp
      stubdata.cpp

      arena.cpp
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(267): error C2485: '__restrict': atributo extendido no reconocido
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(269): error C2485: '__restrict': atributo extendido no reconocido
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(273): error C2485: '__restrict': atributo extendido no reconocido
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(275): error C2485: '__restrict': atributo extendido no reconocido
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(279): error C2485: '__restrict': atributo extendido no reconocido
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/jemalloc.h(281): error C2485: '__restrict': atributo extendido no reconocido
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/internal/jemalloc_internal_macros.h(36): warning C4005: 'restrict': redefinici¢n de macro
      duckdb_build/extension/jemalloc/jemalloc/include\jemalloc/internal/jemalloc_internal_decls.h(87): note: vea la definici¢n anterior de 'restrict'
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.37.32822\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for duckdb
  ```

--- a/.github/workflows/Python.yml
+++ b/.github/workflows/Python.yml
@@ -280,7 +280,7 @@ jobs:
       runs-on: windows-2019
       strategy:
        matrix:
-        python_build: [cp37-*, cp38-*, cp39-*, cp310-*, cp311-*] # , cp312-*] # TODO: figure out windows build issues
+        python_build: [cp37-*, cp38-*, cp39-*, cp310-*, cp311-*, cp312-*]
         isRelease:
           - ${{ startsWith(github.ref, 'refs/tags/v') || github.ref == 'refs/heads/main' }}
         exclude:
@@ -296,7 +296,6 @@ jobs:
 
       env:
         CIBW_BUILD: ${{ matrix.python_build}}
-        SETUPTOOLS_USE_DISTUTILS: 'stdlib'
         TWINE_USERNAME: '__token__'
         DUCKDB_BUILD_UNITY: 1
 

--- a/tools/pythonpkg/setup.py
+++ b/tools/pythonpkg/setup.py
@@ -1,5 +1,6 @@
 #!/usr/bin/env python
 # -*- coding: utf-8 -*-
+import io
 import logging
 import os
 import sys
@@ -7,6 +8,7 @@ import platform
 import multiprocessing.pool
 import traceback
 from glob import glob
+from typing import TextIO
 
 from setuptools import setup, Extension
 from setuptools.command.build_ext import build_ext as _build_ext
@@ -181,6 +183,15 @@ include_directories = [main_include_path, get_pybind_include(), get_pybind_inclu
 if 'BUILD_HTTPFS' in os.environ and 'OPENSSL_ROOT_DIR' in os.environ:
     include_directories += [os.path.join(os.environ['OPENSSL_ROOT_DIR'], 'include')]
 
+
+def exclude_extensions(f: TextIO):
+    files = [x for x in f.read().split('\n') if len(x) > 0]
+    if use_jemalloc:
+        return files
+    else:
+        return [x for x in files if 'jemalloc' not in x]
+
+
 if len(existing_duckdb_dir) == 0:
     # no existing library supplied: compile everything from source
     source_files = main_source_files
@@ -225,14 +236,10 @@ if len(existing_duckdb_dir) == 0:
         # if amalgamation does not exist, we are in a package distribution
         # read the include files, source list and include files from the supplied lists
         with open_utf8('sources.list', 'r') as f:
-            duckdb_sources = [x for x in f.read().split('\n') if len(x) > 0]
-            if not use_jemalloc:
-                duckdb_sources = [x for x in duckdb_sources if 'jemalloc' not in x]
+            duckdb_sources = exclude_extensions(f)
 
         with open_utf8('includes.list', 'r') as f:
-            duckdb_includes = [x for x in f.read().split('\n') if len(x) > 0]
-            if not use_jemalloc:
-                duckdb_includes = [x for x in duckdb_includes if 'jemalloc' not in x]
+            duckdb_includes = exclude_extensions(f)
 
     source_files += duckdb_sources
     include_directories = duckdb_includes + include_directories

--- a/tools/pythonpkg/setup.py
+++ b/tools/pythonpkg/setup.py
@@ -44,6 +44,9 @@ class CompilerLauncherMixin:
         if compiler_launcher:
 
             def spawn_with_compiler_launcher(cmd, **kwargs):
+                if len(' '.join(cmd)) > 32766:
+                    raise Exception("command too long: " + ' '.join(cmd))
+
                 exclude_programs = ("link.exe",)
                 if not cmd[0].endswith(exclude_programs):
                     cmd = [compiler_launcher] + cmd

--- a/tools/pythonpkg/setup.py
+++ b/tools/pythonpkg/setup.py
@@ -42,11 +42,11 @@ class CompilerLauncherMixin:
         compiler_launcher = os.getenv("DISTUTILS_C_COMPILER_LAUNCHER")
         if compiler_launcher:
 
-            def spawn_with_compiler_launcher(cmd):
+            def spawn_with_compiler_launcher(cmd, **kwargs):
                 exclude_programs = ("link.exe",)
                 if not cmd[0].endswith(exclude_programs):
                     cmd = [compiler_launcher] + cmd
-                return original_spawn(cmd)
+                return original_spawn(cmd, **kwargs)
 
             original_spawn = self.compiler.spawn
             self.compiler.spawn = spawn_with_compiler_launcher

--- a/tools/pythonpkg/setup.py
+++ b/tools/pythonpkg/setup.py
@@ -1,13 +1,15 @@
 #!/usr/bin/env python
 # -*- coding: utf-8 -*-
-import io
+import ctypes
 import logging
 import os
 import sys
 import platform
-import multiprocessing.pool
+import sys
 import traceback
+from functools import lru_cache
 from glob import glob
+from os.path import exists
 from typing import TextIO
 
 from setuptools import setup, Extension
@@ -60,6 +62,53 @@ class CompilerLauncherMixin:
 
             original_spawn = self.compiler.spawn
             self.compiler.spawn = spawn_with_compiler_launcher
+
+            # for some reason, the copy of distutils included with setuptools  (instead of the one that was previously
+            # part of the stdlib) pushes the link call over the maximum command line length on windows.
+            # We can fix this by using special windows short paths, of the form C:\Program ~1\...
+
+            def link_with_short_paths(
+                target_desc,
+                objects,
+                *args,
+                **kwargs,
+            ):
+                return original_link(target_desc, [get_short_path(x) for x in objects], *args, **kwargs)
+
+            original_link = self.compiler.link
+            self.compiler.link = link_with_short_paths
+
+
+def get_short_path(long_name: str) -> str:
+    """
+    Gets the short path name of a given long path.
+    http://stackoverflow.com/a/23598461/200291
+    """
+
+    @lru_cache
+    def get_short_path_name_w():
+        from ctypes import wintypes
+
+        kernel32 = ctypes.WinDLL('kernel32', use_last_error=True)
+        _GetShortPathNameW = kernel32.GetShortPathNameW
+        _GetShortPathNameW.argtypes = [wintypes.LPCWSTR, wintypes.LPWSTR, wintypes.DWORD]
+        _GetShortPathNameW.restype = wintypes.DWORD
+        return _GetShortPathNameW
+
+    if platform.system() != 'Windows':
+        return long_name
+
+    assert exists(long_name), long_name
+    gspn = get_short_path_name_w()
+
+    output_buf_size = 0
+    while True:
+        output_buf = ctypes.create_unicode_buffer(output_buf_size)
+        needed = gspn(long_name, output_buf, output_buf_size)
+        if output_buf_size >= needed:
+            return output_buf.value or long_name
+        else:
+            output_buf_size = needed
 
 
 class build_ext(CompilerLauncherMixin, _build_ext):

--- a/tools/pythonpkg/setup.py
+++ b/tools/pythonpkg/setup.py
@@ -81,36 +81,6 @@ unity_build = 0
 if 'DUCKDB_BUILD_UNITY' in os.environ:
     unity_build = 16
 
-
-def parallel_cpp_compile(
-    self,
-    sources,
-    output_dir=None,
-    macros=None,
-    include_dirs=None,
-    debug=0,
-    extra_preargs=None,
-    extra_postargs=None,
-    depends=None,
-):
-    # Copied from distutils.ccompiler.CCompiler
-    macros, objects, extra_postargs, pp_opts, build = self._setup_compile(
-        output_dir, macros, include_dirs, sources, depends, extra_postargs
-    )
-
-    cc_args = self._get_cc_args(pp_opts, debug, extra_preargs)
-
-    def _single_compile(obj):
-        try:
-            src, ext = build[obj]
-        except KeyError:
-            return
-        self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
-
-    list(multiprocessing.pool.ThreadPool(multiprocessing.cpu_count()).imap(_single_compile, objects))
-    return objects
-
-
 # speed up compilation with: -j = cpu_number() on non Windows machines
 if os.name != 'nt' and os.environ.get('DUCKDB_DISABLE_PARALLEL_COMPILE', '') != '1':
     try:
@@ -299,12 +269,6 @@ else:
         language='c++',
         define_macros=define_macros,
     )
-
-# Only include pytest-runner in setup_requires if we're invoking tests
-if {'pytest', 'test', 'ptr'}.intersection(sys.argv):
-    setup_requires = ['pytest-runner']
-else:
-    setup_requires = []
 
 
 # data files need to be formatted as [(directory, [files...]), (directory2, [files...])]

--- a/tools/pythonpkg/setup.py
+++ b/tools/pythonpkg/setup.py
@@ -5,6 +5,7 @@ import os
 import sys
 import platform
 import multiprocessing.pool
+import traceback
 from glob import glob
 
 from setuptools import setup, Extension
@@ -46,7 +47,11 @@ class CompilerLauncherMixin:
                 exclude_programs = ("link.exe",)
                 if not cmd[0].endswith(exclude_programs):
                     cmd = [compiler_launcher] + cmd
-                return original_spawn(cmd, **kwargs)
+                try:
+                    return original_spawn(cmd, **kwargs)
+                except Exception:
+                    traceback.print_exc()
+                    raise
 
             original_spawn = self.compiler.spawn
             self.compiler.spawn = spawn_with_compiler_launcher

--- a/tools/pythonpkg/setup.py
+++ b/tools/pythonpkg/setup.py
@@ -154,22 +154,22 @@ for i in range(len(sys.argv)):
     else:
         new_sys_args.append(sys.argv[i])
 sys.argv = new_sys_args
-toolchain_args.append('-DDUCKDB_PYTHON_LIB_NAME=' + lib_name)
+define_macros = [('DUCKDB_PYTHON_LIB_NAME', lib_name)]
 
 if platform.system() == 'Darwin':
     toolchain_args.extend(['-stdlib=libc++', '-mmacosx-version-min=10.7'])
 
 if platform.system() == 'Windows':
-    toolchain_args.extend(['-DDUCKDB_BUILD_LIBRARY', '-DWIN32'])
+    define_macros.extend([('DUCKDB_BUILD_LIBRARY', None), ('WIN32', None)])
 
 if 'BUILD_HTTPFS' in os.environ:
     libraries += ['crypto', 'ssl']
     extensions += ['httpfs']
 
 for ext in extensions:
-    toolchain_args.extend(['-DDUCKDB_EXTENSION_{}_LINKED'.format(ext.upper())])
+    define_macros.append(('DUCKDB_EXTENSION_{}_LINKED'.format(ext.upper()), None))
 
-toolchain_args.extend(['-DDUCKDB_EXTENSION_AUTOLOAD_DEFAULT=1', '-DDUCKDB_EXTENSION_AUTOINSTALL_DEFAULT=1'])
+define_macros.extend([('DUCKDB_EXTENSION_AUTOLOAD_DEFAULT', '1'), ('DUCKDB_EXTENSION_AUTOINSTALL_DEFAULT', '1')])
 
 linker_args = toolchain_args
 if platform.system() == 'Windows':
@@ -267,6 +267,7 @@ if len(existing_duckdb_dir) == 0:
         extra_link_args=linker_args,
         libraries=libraries,
         language='c++',
+        define_macros=define_macros,
     )
 else:
     sys.path.append(os.path.join(script_path, '..', '..', 'scripts'))
@@ -288,6 +289,7 @@ else:
         libraries=libnames,
         library_dirs=library_dirs,
         language='c++',
+        define_macros=define_macros,
     )
 
 # Only include pytest-runner in setup_requires if we're invoking tests

